lib/filterx/func-str.c

```diff
@@ -0,0 +1,387 @@
+  gsize str_len;
+} FilterXStringWithCache;
+
+typedef gboolean (*FilterXExprAffixProcessFunc)(const gchar *haystack, gsize haystack_len, const gchar *needle,
+                                                gsize needle_len);
+typedef struct _FilterXExprAffix
+{
+  FilterXFunction super;
+  gboolean ignore_case;
+  FilterXStringWithCache *haystack;
+  struct
+  {
+    FilterXExpr *expr;
+    GPtrArray *needle_objects_with_cache;
+  } needle;
+  FilterXExprAffixProcessFunc process;
+} FilterXExprAffix;
+
+static FilterXStringWithCache *
+filterx_string_with_cache_new(void)
+{
+  FilterXStringWithCache *self = g_new0(FilterXStringWithCache, 1);
+  return self;
+}
+
+static gboolean
+_do_casefold(const gchar *src, gsize *len, gchar **dst)
+{
+  if(*len > G_MAXSSIZE)
+    return FALSE;
+
+  *dst = g_utf8_casefold(src, (gssize) *len);
+  *len = g_utf8_strlen(*dst, -1);
+  return TRUE;
+}
+
+static gboolean
+_filterx_string_with_cache_eval_uncached(FilterXStringWithCache *self, gboolean ignore_case, const gchar **dst,
+                                         gsize *len)
+{
+  self->obj = filterx_expr_eval(self->expr);
+  if (!self->obj)
+    {
+      filterx_eval_push_error_info("failed to evaluate expression", self->expr,  g_strdup_printf("invalid expression"), TRUE);
+      return FALSE;
+    }
+  if (!filterx_object_extract_string(self->obj, &self->borrowed_str_value, &self->str_len))
+    {
+      filterx_eval_push_error_info("failed to extract string from object, it must be a string", self->expr,
+                                   g_strdup_printf("got %s instead", self->obj->type->name), TRUE);
+      filterx_object_unref(self->obj);
+      return FALSE;
+    }
+  if (ignore_case)
+    {
+      if(!_do_casefold(self->borrowed_str_value, &self->str_len, &self->owned_str_value))
+        {
+          filterx_eval_push_error_info("failed to casefold, string too long", self->expr,
+                                       g_strdup_printf("length: %ld, max supported length: %ld ", self->str_len, G_MAXSSIZE), TRUE);
+          filterx_object_unref(self->obj);
+          return FALSE;
+        }
+      *dst = self->owned_str_value;
+      *len = self->str_len;
+      return TRUE;
+    }
+  *dst = self->borrowed_str_value;
+  *len = self->str_len;
+  return TRUE;
+}
+
+static gboolean
+filterx_string_with_cache_get_string_value(FilterXStringWithCache *self, gboolean ignore_case, const gchar **dst,
+                                           gsize *len)
+{
+  if (ignore_case && self->owned_str_value)
+    {
+      *dst = self->owned_str_value;
+      *len = self->str_len;
+      return TRUE;
+    }
+  else if (!ignore_case && self->borrowed_str_value)
+    {
+      *dst = self->borrowed_str_value;
+      *len = self->str_len;
+      return TRUE;
+    }
+
+  return _filterx_string_with_cache_eval_uncached(self, ignore_case, dst, len);
+}
+
+static gboolean
+_cache_string_literal(FilterXExpr *expr, const gchar *expression_name, gboolean ignore_case,
```

Comment:
I am finding the different prefixes to these functions difficult to read, as I am trying to find out why some of them has a _filterx_ prefix while others do not.

The many "cache" related functions and the FilterXStringWithCache struct makes it difficult to judge which class the specific method is related to.

As I see most of this function should be delegated to FilterXStringWithCache as it directly changes members of that struct.